copy/copy.go

```diff
@@ -496,27 +486,29 @@ func (ic *imageCopier) copyLayers(ctx context.Context) error {
 			cld.destInfo, cld.diffID, cld.err = ic.copyLayer(ctx, srcLayer, bar)
 		}
 		data[index] = cld
-		if bar != nil {
-			bar.SetTotal(srcLayer.Size, true)
-		}
+		bar.SetTotal(srcLayer.Size, true)
 	}
 
-	progressBars := make([]*mpb.Bar, numLayers)
-	for i, srcInfo := range srcInfos {
-		progressBars[i] = ic.c.createProgressBar(srcInfo, shortDigest(srcInfo.Digest), "blob")
-	}
+	func() { // A scope for defer
+		progressPool, progressCleanup := ic.c.newProgressPool(ctx)
+		defer progressCleanup()
 
-	for i, srcLayer := range srcInfos {
-		copySemaphore.Acquire(ctx, 1)
-		go copyLayerHelper(i, srcLayer, progressBars[i])
-	}
+		progressBars := make([]*mpb.Bar, numLayers)
+		for i, srcInfo := range srcInfos {
+			progressBars[i] = ic.c.createProgressBar(progressPool, srcInfo, "blob")
+		}
 
-	destInfos := make([]types.BlobInfo, numLayers)
-	diffIDs := make([]digest.Digest, numLayers)
+		for i, srcLayer := range srcInfos {
+			copySemaphore.Acquire(ctx, 1)
+			go copyLayerHelper(i, srcLayer, progressBars[i])
+		}
 
-	// Wait for all layers to be copied
-	copyGroup.Wait()
+		// Wait for all layers to be copied
+		copyGroup.Wait()
+	}()
 
+	destInfos := make([]types.BlobInfo, numLayers)
+	diffIDs := make([]digest.Digest, numLayers)
```

Comment:
Nit: the intention behind moving those before `copyGroup.Wait()` was that the allocation can happen while we're still waiting for the copy-goroutines to finish. Maybe, I am missing something but is the `defer progressCleanup()` technically required or just means to make the code more readable?

I *think* that the following would be slightly faster:
```golang
	destInfos := make([]types.BlobInfo, numLayers)
	diffIDs := make([]digest.Digest, numLayers)
        copyGroup.Wait()
        progressCleanup()
```

Reply:
> Nit: the intention behind moving those before `copyGroup.Wait()` was that the allocation can happen while we're still waiting for the copy-goroutines to finish.

That’s thoughtful and clever, but we are going to be blocked (many times) on disk and possibly network, is moving a few memory allocations to run concurrently with that at all noticeable, and worth the readability impact? In this case it would mean two extra return values from the `func() { // A scope for defer`, or declaring the variables (as `nil` pointers) before that function starts, the function, adding two more links and stretching the scope of the variables much larger.

To me, computers seem fast enough that I’d much rather have short easy-to read code than long clever code, unless it is a known tight loop critical for the product (which wouldn’t be written in Go at all, most likely).


> Maybe, I am missing something but is the `defer progressCleanup()` technically required or just means to make the code more readable?

It also ensures that the progress bar is properly cleaned up in case of a `panic` (right now), which is admittedly pretty unlikely to happen — and also if any future updates add a `return err` somewhere within the scope of `progressPool`, which is much more likely (it already happens in `copyConfig`). So, it contributes to longer-term reliability.

---

Up to you, I can change it back if you prefer the speed-up.

Reply:
> Up to you, I can change it back if you prefer the speed-up.

I am fine with this code. The maintainability overrules the few cycles we may safe.